src/GitHub.VisualStudio/GitHubPanePackage.cs

```diff
@@ -0,0 +1,40 @@
+﻿using System;
+using System.Runtime.InteropServices;
+using GitHub.VisualStudio.UI;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.ComponentModelHost;
+
+namespace GitHub.VisualStudio
+{
+    /// <summary>
+    /// This is the host package for the <see cref="GitHubPane"/> tool window.
+    /// </summary>
+    /// <remarks>
+    /// We auto-load this package before its tool window is activated to ensure that MEF has
+    /// been initialized and the tool window won't be blamed for degrading startup performance.
+    /// See: https://github.com/github/VisualStudio/issues/1550
+    /// </remarks>
+    [PackageRegistration(UseManagedResourcesOnly = true)]
+    [Guid(Guids.InitializeMefPackageId)]
+    // This package must auto-load before its tool window is activated on startup.
+    // We're using the GitSccProvider UI context because this is triggered before
+    // tool windows are activated and we already use it elsewhere (the containing
+    // assembly would have been loaded anyway).
```

Comment:
VS Shell will already ensure your package is loaded before the tool window is activated so you don't have to do it here. 

This auto load ensures that your package is loaded for every startup session regardless of your tool window is open slowing down startup for all users instead of just those with the toolwindow visible.

Reply:
> VS Shell will already ensure your package is loaded before the tool window is activated so you don't have to do it here.

If we allow that to happen, our package gets blamed for MEF initialization (which can take 10 seconds 😨).

> This auto load ensures that your package is loaded for every startup session regardless of your tool window is open slowing down startup for all users instead of just those with the toolwindow visible.

We need to auto-load on this context anyway in order to install a custom `UIContext`. That's why I wasn't too concerned about loading this assembly.

Reply:
As you pointed, the cost is still there just becomes part of SCC package which is unique in this case since it is the one that causes your auto load unlike other scenarios. At the end of the day startup is still impacted by 10 seconds which is the important part. I have to mention that SCC hiding cost of auto loads from the UI context it sets is a known issue that we will fix soon btw.

As for your auto load reason, would you be able to use a rule based UI context to activate your UI context instead of auto loading a package?

Reply:
> I have to mention that SCC hiding cost of auto loads from the UI context it sets is a known issue that we will fix soon btw.

Thanks, this is good to know.

> As for your auto load reason, would you be able to use a rule based UI context to activate your UI context instead of auto loading a package?

Our UI context is a subset of the `GitSccProvider` context. Is there a way to implement a custom UI context that doesn't involve loading a package?

Am I correct in thinking that a rule based UI context is a composite of existing UI context (you can't use them to create a completely custom context)?

Reply:
What are the conditions that you set the UI context in? If it is GIT SCC context and presence of your package, you can just define a rule based UI contexts that is only based on GIT SCC context. Since the rule based UI context will only be added when your pkgdef is deployed it will only be active when GIT SCC is active and user has your extension installed.

Reply:
> What are the conditions that you set the UI context in?

Here is pseudo code for what we need:
If in `GitSccProvider` context `&&` `IGitExt.ActiveRepositories.Count > 0`

We're pretty much always in the `GitSccProvider` context, because it's active even when there is no loaded solution but the previously loaded solution was in a Git repo. 😕 We therefore check the `IGitExt` service to find out if we're really in the context of an active repo.

We have made efforts to only load when absolutely necessary! 😄